File: src/main/java/com/backend/service/DTO/OffreDTO.java
package com.backend.service.DTO;

import com.backend.modele.Offre;
import lombok.*;

import java.time.LocalDate;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class OffreDTO {

    private Long id;
    private AuthResponseDTO authResponseDTO;
    private String titre;
    private String description;
    private LocalDate date_debut;
    private LocalDate date_fin;
    private int annee;
    private ProgrammeDTO progEtude;
    private String lieuStage;
    private String remuneration;
    private String horaire;
    private Integer dureeHebdomadaire;
    private String responsabilitesEtudiant;
    private String responsabilitesEmployeur;
    private String responsabiliteCollege;
    private String objectifs;
    private LocalDate dateLimite;
    private String messageRefus;
    private String statutApprouve;
    private EmployeurDTO employeurDTO;

    public OffreDTO toDTO(Offre offre) {
        OffreDTO dto = new OffreDTO();
        dto.setId(offre.getId());
        dto.setAuthResponseDTO(this.authResponseDTO);
        dto.setTitre(offre.getTitre());
        dto.setDescription(offre.getDescription());
        dto.setDate_debut(offre.getDate_debut());
        dto.setDate_fin(offre.getDate_fin());
        dto.setAnnee(offre.getAnnee());
        dto.setProgEtude(ProgrammeDTO.toDTO(offre.getProgEtude()));
        dto.setLieuStage(offre.getLieuStage());
        dto.setRemuneration(offre.getRemuneration());
        dto.setHoraire(offre.getHoraire());
        dto.setDureeHebdomadaire(offre.getDureeHebdomadaire());
        dto.setResponsabilitesEtudiant(offre.getResponsabilitesEtudiant());
        dto.setResponsabilitesEmployeur(offre.getResponsabilitesEmployeur());
        dto.setResponsabiliteCollege(offre.getResponsabilitesCollege());
        dto.setObjectifs(offre.getObjectifs());
        dto.setDateLimite(offre.getDateLimite());
        dto.setMessageRefus(offre.getMessageRefus());
        dto.setStatutApprouve(offre.getStatutApprouve() != null ? offre.getStatutApprouve().toString() : "ATTENTE");
        dto.setEmployeurDTO(new EmployeurDTO().toDTO(offre.getEmployeur()));
        return dto;
    }
}

File: frontend_mobile/lib/models/offre_dto.dart
class OffreDTO {
  final int? id;
  final String? titre;
  final String? description;
  final String? date_debut;
  final String? date_fin;
  final String? progEtude;
  final String? lieuStage;
  final String? remuneration;
  final String? dateLimite;
  final String? nomEntreprise;

  OffreDTO({
    this.id,
    this.titre,
    this.description,
    this.date_debut,
    this.date_fin,
    this.progEtude,
    this.lieuStage,
    this.remuneration,
    this.dateLimite,
    this.nomEntreprise,
  });

  factory OffreDTO.fromJson(Map<String, dynamic> json) {
    final idValue = json['id'];
    final id = idValue is int
        ? idValue
        : (idValue is double
              ? idValue.toInt()
              : (idValue is String ? int.tryParse(idValue) : null));

    // nomEntreprise provient de l'objet employeurDTO.nomEntreprise
    String? nomEntreprise;
    try {
      final emp = json['employeurDTO'];
      if (emp is Map<String, dynamic>) {
        nomEntreprise = emp['nomEntreprise']?.toString();
      }
    } catch (_) {
      nomEntreprise = null;
    }

    return OffreDTO(
      id: id,
      titre: json['titre']?.toString(),
      description: json['description']?.toString(),
      date_debut: json['date_debut']?.toString(),
      date_fin: json['date_fin']?.toString(),
      progEtude: json['progEtude']?.toString(),
      lieuStage: json['lieuStage']?.toString(),
      remuneration: json['remuneration']?.toString(),
      dateLimite: json['dateLimite']?.toString(),
      nomEntreprise: nomEntreprise,
    );
  }
}
